### DISCUSSION TITLE
Tapping into the message pump?

### DISCUSSION THREAD

**Discussion by AsafMahlev** [AvaloniaUI/Avalonia#8650] [Q&A] [answered]
I'm working on a clipboard manager program, and to listen to clipboards in windows, you use this method to get change notifications: https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-addclipboardformatlistener?redirectedfrom=MSDN

What it does is push a message into the message pump.

Now, when developing an app in Avalonia, is there a way to access the pump? In winforms you can inherit WndProc, but I don't think I've seen anything like this here.

What would be a good alternative?

**Comment by maxkatz6** [comment]
Use GetWindowLongPtr64 and SetWindowLongPtr64. 
Something like this should work:

```csharp
hWndProcHook = GetWindowLongPtr(handle, GetWindowLongPtr_nIndex.GWLP_WNDPROC);
var newLong = Marshal.GetFunctionPointerForDelegate(_fnWndProcHook);
SetWindowLongPtr(handle, GetWindowLongPtr_nIndex.GWLP_WNDPROC, newLong);

private unsafe LRESULT WndProcHook(HWND hwnd, uint code, WPARAM wParam, LPARAM lParam)
{
     return CallWindowProc(_hWndProcHook, hwnd, code, wParam, lParam);
}
```

Where definitions and constants you can find in MS documentation. Or use [C#/Win32](https://github.com/microsoft/CsWin32) generator.

**Reply by AsafMahlev** [reply]
Thanks! But the handle to the window in Win32Platform is `internal`.

Is there a proper way to access it? (other than reflection?)

**Reply by kekekeks** [reply]
`window.TryGetPlatformHandle()?.Handle`

**Reply by AsafMahlev** [reply]
Using the main window makes my program crash after a few seconds.
Using the Win32Platform window by reflection works though.

**Reply by AsafMahlev** [reply]
Never mind, that window crashes too when resizing or just sometimes.

My method is:

```c#
    private static LRESULT WndProcHook(HWND hwnd, uint code, WPARAM wParam, LPARAM lParam)
    {
        if (code == WM_CLIPBOARD_UPDATE)
        {
            ClipboardNewEvent?.Invoke(null, EventArgs.Empty);
            return (LRESULT)0;
        }
        return WindowLongPtrHelper.CallWindowProc(_originalWndproc, hwnd, code, wParam, lParam);
    }
```

Doesn't seem wrong.

**Reply by kekekeks** [reply]
Make sure that you pin the delegate

**Reply by swety2003** [reply]
I tried with [CsWin32](https://github.com/Microsoft/CsWin32) and my code is below:
```
using Avalonia.Controls;
using System.Runtime.InteropServices;
using Windows.Win32.Foundation;
using Windows.Win32.UI.WindowsAndMessaging;
using static Windows.Win32.PInvoke;

namespace AvaloniaApplication1.Views
{
    public partial class MainWindow : Window
    {
        private nint hWndProcHook;

        public MainWindow()
        {
            InitializeComponent();
        }

        protected override void OnLoaded(RoutedEventArgs e)
        {
            base.OnLoaded(e);
            var handle = new HWND(TryGetPlatformHandle().Handle);
            var customWndProc = new WndProc(WndProcHook);

            hWndProcHook = GetWindowLongPtr(handle, WINDOW_LONG_PTR_INDEX.GWLP_WNDPROC);
            var newLong = Marshal.GetFunctionPointerForDelegate(customWndProc);
            SetWindowLongPtr(handle, WINDOW_LONG_PTR_INDEX.GWLP_WNDPROC, (nint)newLong);

        }
        private LRESULT WndProcHook(HWND hwnd, uint code, WPARAM wParam, LPARAM lParam)
        {
            //do something
            //...
            return Native.CallWindowProc(hWndProcHook, hwnd, code, wParam, lParam);
        }
        private delegate LRESULT WndProc(HWND hwnd, uint msg, WPARAM wParam, LPARAM lParam);

    }

    internal static class Native
    {
        [DllImport("User32.dll", SetLastError = true)]
        internal static extern LRESULT CallWindowProc(nint lpPrevWndFunc, HWND hWnd, uint Msg, WPARAM wParam, LPARAM lParam);
    }

}
```
However, it throws System.ExecutionEngineException after a few seconds.
Anyone can help me?

**Reply by maxkatz6** [reply]
You forgot to pin the delegate. "customWndProc" object need to have a lifetime bigger than just a method. Otherwise GC will garbage collect it.

**Comment by grokys** [comment, marked as answer]
For anyone finding this issue in future, 11.1 will have an API to do this:

```
Win32Properties.AddWndProcHookCallback(window, WndProcHook);
```
